### PR TITLE
Gale Shapley Algorithm

### PR DESCRIPTION
**Pull request by akashvshroff**
Implementation of a Nobel prize-winning algorithm that determines a stable matching in a bipartite graph.

### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by mateuszz0000** [comment]
Please `flake8 --ignore=E203,W503 --max-complexity=25 --max-line-length=120 --statistics --count .` on your local machine and fix errors.

File: graphs/gale_shapley_bigraph.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    '''
+    Finds the stable match in any bipartite graph, i.e a pairing where no 2 objects prefer each other over their partner.
+    Here marriage is used to make the variable names easier and so the algorithm can be intuitively understood.
+    The function accepts the preferences of the men and women (where both are named from 0 to n-1) and returns a list where index
+    position corresponds to the man and value at the index is the woman he is marrying.
+    E.g:
+    n = 4
+    men_preferences = [[0, 1, 3, 2], [0, 2, 3, 1], [1, 0, 2, 3], [0, 3, 1, 2]]
+    women_preferences = [[3, 1, 2, 0], [3, 1, 0, 2], [0, 3, 1, 2], [1, 0, 3, 2]]
+    >>>print(stable_matching(n,men_preferences,women_preferences))
```
Comment (cclauss) [inline graphs/gale_shapley_bigraph.py:11]:
```suggestion
    >>> print(stable_matching(n,men_preferences,women_preferences))
```

**Comment by akashvshroff** [comment]
> Please `flake8 --ignore=E203,W503 --max-complexity=25 --max-line-length=120 --statistics --count . `on your local machine and fix errors.

I think it should be alright now? Thanks.

**Comment by wuyudi** [comment]
I read the log, it said `W291 trailing whitespace` which means there are extra white sapces. You can use the format function in your editor to fix it.

**Comment by akashvshroff** [comment]
> I read the log, it said `W291 trailing whitespace` which means there are extra white spaces. You can use the format function in your editor to fix it.

When I do it, it says 0 and indicates there are no errors - did you try it on the code as part of my latest commit? Thanks.

**Comment by cclauss** [comment]
psf/black and other formatters might be reluctant to modify text that is inside a triple quoted string but flake8 will still flag that those lines as having trailing whitespace.  If that is the case then manual edits may be required.

**Comment by cclauss** [comment]
I know this is a bit of a chore but...  I do not think that we should base an algorithm on the premise that `Marriages are heterosexual` as it will not be viewed as politically correct in some circles.  Could we please shift to `donors` and `recipients` instead.  I hope that you understand why this would be helpful for use building a diversified community of contributors.  Thanks.

**Comment by akashvshroff** [comment]
> I know this is a bit of a chore but... I do not think that we should base an algorithm on the premise that `Marriages are heterosexual` as it will not be viewed as politically correct in some circles. Could we please shift to organ `donors` and `recipients` instead. I hope that you understand why this would be helpful for us building a diversified community of contributors. Thanks.

I've made the necessary changes and manually ensured that there is no trailing whitespace in the program, I hope it is alright now! Thanks.

File: graphs/gale_shapley_bigraph.py [hunk 2]
```diff
@@ -1,43 +1,37 @@
...
-    Finds the stable match in any bipartite graph, i.e a pairing where no 2 
+    Finds the stable match in any bipartite graph, i.e a pairing where no 2
     objects prefer each other over their partner.
-    Here marriage is used to make the variable names easier and so the algorithm 
-    can be intuitively understood.
-    The function accepts the preferences of the men and women (where both are 
-    named from 0 to n-1) and returns a list where index position corresponds 
-    to the man and value at the index is the woman he is marrying.
+    The function accepts the preferences of the donors and recipients (where
+    both arecnamed from 0 to n-1) and returns a list where the index position
```
Comment (cclauss) [inline graphs/gale_shapley_bigraph.py:6]:
```suggestion
    both are assigned numbers from 0 to n-1) and returns a list where the index position
```

File: graphs/gale_shapley_bigraph.py [hunk 3]
```diff
@@ -1,43 +1,37 @@
...
+    corresponds to the donor and value at the index is the recipient (of the organ).
```
Comment (cclauss) [inline graphs/gale_shapley_bigraph.py:7]:
```suggestion
    corresponds to the donor and value at the index is the organ recipient.
```

File: graphs/gale_shapley_bigraph.py [hunk 4]
```diff
@@ -1,43 +1,37 @@
...
-        woman = his_preferences[num_proposals[man]]
-        num_proposals[man] += 1
-        her_preferences = women_pref[woman]
-        husb = woman_spouse[woman]
-        if husb is not None:
-            if her_preferences.index(husb) > her_preferences.index(man):
-                woman_spouse[woman], man_spouse[man] = man, woman
-                unmarried_men.append(husb)
-                unmarried_men.remove(man)
+    undonated_donors = [i for i in range(n)]
```
Comment (cclauss) [inline graphs/gale_shapley_bigraph.py:18]:
```suggestion
    unmatched_donors = [i for i in range(n)]
```

**Comment by cclauss** [comment]
./graphs/gale_shapley_bigraph.py:30:89: E501 line too long (90 > 88 characters)

**Comment by akashvshroff** [comment]
I hope I've edited everything as required!

**Pull request merged**